File: src/main/java/com/chatbi/agent/skills/SkillLoader.java
package com.chatbi.agent.skills;

import lombok.extern.slf4j.Slf4j;

import java.io.IOException;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Stream;

/**
 * SkillLoader - 技能加载器
 * 
 * 从标准目录中发现并加载技能。
 * 
 * 搜索路径（按优先级顺序）：
 * 1. .claude/skills (项目级)
 * 2. .minion/skills (项目级)
 * 3. ~/.claude/skills (用户级)
 * 4. ~/.minion/skills (用户级)
 * 
 * 项目级技能覆盖同名的用户级技能。
 */
@Slf4j
public class SkillLoader {
    
    /**
     * 默认技能目录名称
     */
    private static final List<String> SKILL_DIRS = Arrays.asList(
        ".claude/skills",
        ".minion/skills",
        // 项目内置技能目录
        "src/main/java/com/chatbi/agent/skills/skills"
    );
    
    /**
     * 技能文件名
     */
    private static final String SKILL_FILE = "SKILL.md";
    
    /**
     * 项目根目录
     */
    private final Path projectRoot;
    
    /**
     * 用户主目录
     */
    private final Path homeDir;
    
    /**
     * 额外的技能搜索路径
     */
    private final List<Path> additionalPaths = new ArrayList<>();
    
    // ==================== 构造函数 ====================
    
    public SkillLoader() {
        this(Paths.get(System.getProperty("user.dir")));
    }
    
    public SkillLoader(Path projectRoot) {
        this.projectRoot = projectRoot != null ? projectRoot : Paths.get(System.getProperty("user.dir"));
        this.homeDir = Paths.get(System.getProperty("user.home"));
    }
    
    // ==================== 配置方法 ====================
    
    /**
     * 添加额外的技能搜索路径
     * 
     * @param path 搜索路径
     * @return this（用于链式调用）
     */
    public SkillLoader addSearchPath(Path path) {
        if (path != null) {
            additionalPaths.add(path);
        }
        return this;
    }
    
    /**
     * 添加额外的技能搜索路径
     * 
     * @param path 搜索路径字符串
     * @return this
     */
    public SkillLoader addSearchPath(String path) {
        if (path != null && !path.isEmpty()) {
            additionalPaths.add(Paths.get(path));
        }
        return this;
    }
    
    // ==================== 搜索路径 ====================
    
    /**
     * 获取所有搜索路径
     * 
     * @return (路径, 位置类型) 元组列表
     */
    public List<SearchPath> getSearchPaths() {
        List<SearchPath> paths = new ArrayList<>();
        
        // 项目级路径（更高优先级）
        for (String skillDir : SKILL_DIRS) {
            Path projectPath = projectRoot.resolve(skillDir);
            paths.add(new SearchPath(projectPath, "project"));
        }
        
        // 用户级路径（较低优先级）
        for (String skillDir : SKILL_DIRS) {
            Path userPath = homeDir.resolve(skillDir);
            paths.add(new SearchPath(userPath, "user"));
        }
        
        // 额外路径
        for (Path additionalPath : additionalPaths) {
            paths.add(new SearchPath(additionalPath, "managed"));
        }
        
        return paths;
    }
    
    /**
     * 搜索路径记录
     */
    public record SearchPath(Path path, String location) {}
    
    // ==================== 发现技能 ====================
    
    /**
     * 在技能目录中发现所有技能
     * 
     * @param skillsDir 包含技能子目录的目录
     * @return SKILL.md文件路径列表
     */
    public List<Path> discoverSkills(Path skillsDir) {
        List<Path> skillFiles = new ArrayList<>();
        
        if (!Files.exists(skillsDir) || !Files.isDirectory(skillsDir)) {
            return skillFiles;
        }
        
        try (Stream<Path> items = Files.list(skillsDir)) {
            items.filter(Files::isDirectory).forEach(item -> {
                // 检查目录中是否有SKILL.md
                Path skillMd = item.resolve(SKILL_FILE);
                
                if (Files.exists(skillMd)) {
                    skillFiles.add(skillMd);
                } else {
                    // 检查嵌套子目录
                    try (Stream<Path> nestedItems = Files.list(item)) {
                        nestedItems.filter(Files::isDirectory).forEach(nestedItem -> {
                            Path nestedSkillMd = nestedItem.resolve(SKILL_FILE);
                            if (Files.exists(nestedSkillMd)) {
                                skillFiles.add(nestedSkillMd);
                            }
                        });
                    } catch (IOException e) {
                        log.debug("Error listing nested directory: {}", item, e);
                    }
                }
            });
        } catch (IOException e) {
            log.debug("Error listing skills directory: {}", skillsDir, e);
        }
        
        return skillFiles;
    }
    
    // ==================== 加载技能 ====================
    
    /**
     * 从SKILL.md文件加载单个技能
     * 
     * @param skillMdPath SKILL.md文件路径
     * @param location 位置类型
     * @return Skill实例，加载失败返回null
     */
    public Skill loadSkill(Path skillMdPath, String location) {
        try {
            Skill skill = Skill.fromSkillMd(skillMdPath, location);
            if (skill != null) {
                log.debug("Loaded skill: {} from {}", skill.getName(), skillMdPath);
            } else {
                log.warn("Failed to parse skill: {}", skillMdPath);
            }
            return skill;
        } catch (Exception e) {
            log.error("Error loading skill from {}: {}", skillMdPath, e.getMessage());
            return null;
        }
    }
    
    /**
     * 加载所有搜索路径中的技能
     * 
     * @return 包含所有技能的SkillRegistry
     */
    public SkillRegistry loadAll() {
        return loadAll(null);
    }
    
    /**
     * 加载所有搜索路径中的技能到指定注册表
     * 
     * @param registry 目标注册表，为null时使用全局实例
     * @return 包含所有技能的SkillRegistry
     */
    public SkillRegistry loadAll(SkillRegistry registry) {
        if (registry == null) {
            registry = SkillRegistry.getGlobalInstance();
        }
        
        int totalLoaded = 0;
        
        for (SearchPath searchPath : getSearchPaths()) {
            List<Path> skillFiles = discoverSkills(searchPath.path());
            
            for (Path skillMdPath : skillFiles) {
                Skill skill = loadSkill(skillMdPath, searchPath.location());
                if (skill != null) {
                    boolean registered = registry.register(skill);
                    if (registered) {
                        totalLoaded++;
                    }
                }
            }
        }
        
        log.info("Loaded {} skills from all search paths", totalLoaded);
        return registry;
    }
    
    /**
     * 重新加载所有技能（先清空现有技能）
     * 
     * @return 重新加载后的SkillRegistry
     */
    public SkillRegistry reload() {
        return reload(null);
    }
    
    /**
     * 重新加载所有技能到指定注册表
     * 
     * @param registry 目标注册表
     * @return 重新加载后的SkillRegistry
     */
    public SkillRegistry reload(SkillRegistry registry) {
        if (registry == null) {
            registry = SkillRegistry.getGlobalInstance();
        }
        
        registry.clear();
        return loadAll(registry);
    }
    
    /**
     * 从指定目录加载技能
     * 
     * @param skillsDir 技能目录
     * @param location 位置类型
     * @param registry 目标注册表
     * @return 加载的技能数量
     */
    public int loadFromDirectory(Path skillsDir, String location, SkillRegistry registry) {
        if (registry == null) {
            registry = SkillRegistry.getGlobalInstance();
        }
        
        List<Path> skillFiles = discoverSkills(skillsDir);
        int count = 0;
        
        for (Path skillMdPath : skillFiles) {
            Skill skill = loadSkill(skillMdPath, location);
            if (skill != null && registry.register(skill)) {
                count++;
            }
        }
        
        log.info("Loaded {} skills from {}", count, skillsDir);
        return count;
    }
    
    // ==================== 便捷方法 ====================
    
    /**
     * 便捷函数：加载所有技能
     * 
     * @return SkillRegistry
     */
    public static SkillRegistry loadSkills() {
        return loadSkills(null);
    }
    
    /**
     * 便捷函数：从指定项目根目录加载所有技能
     * 
     * @param projectRoot 项目根目录
     * @return SkillRegistry
     */
    public static SkillRegistry loadSkills(Path projectRoot) {
        SkillLoader loader = new SkillLoader(projectRoot);
        return loader.loadAll();
    }
    
    /**
     * 获取所有可用技能列表
     * 
     * @return 技能列表
     */
    public static List<Skill> getAvailableSkills() {
        SkillRegistry registry = SkillRegistry.getGlobalInstance();
        
        if (registry.isEmpty()) {
            loadSkills();
        }
        
        return registry.listAll();
    }
    
    @Override
    public String toString() {
        return String.format("SkillLoader[projectRoot=%s, searchPaths=%d]", 
            projectRoot, getSearchPaths().size());
    }
}

File: minion/minion/skills/skill_loader.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Skill Loader - discovers and loads skills from standard directories.
"""

import os
from pathlib import Path
from typing import List, Optional
import logging

from .skill import Skill
from .skill_registry import SkillRegistry, get_skill_registry

logger = logging.getLogger(__name__)


class SkillLoader:
    """
    从标准目录中查找并加载技能

    Search paths (in priority order):
    1. .claude/skills (project-level)
    2. .minion/skills (project-level)
    3. ~/.claude/skills (user-level)
    4. ~/.minion/skills (user-level)

    Project-level skills override user-level skills with the same name.
    """

    # Default skill directory names
    SKILL_DIRS = [
        ".claude/skills",
        ".minion/skills",
    ]

    SKILL_FILE = "SKILL.md"

    def __init__(self, project_root: Optional[Path] = None):
        """
        Initialize the skill loader.

        Args:
            project_root: Root directory of the project. Defaults to current directory.
        """
        self.project_root = Path(project_root) if project_root else Path.cwd()
        self.home_dir = Path.home()

    def get_search_paths(self) -> List[tuple[Path, str]]:
        """
        从项目目录和用户目录中查找技能，并保存技能具体目录

        Returns:
            List of (path, location_type) tuples
        """
        paths = []

        # Project-level paths (higher priority)
        for skill_dir in self.SKILL_DIRS:
            project_path = self.project_root / skill_dir
            paths.append((project_path, "project"))

        # User-level paths (lower priority)
        for skill_dir in self.SKILL_DIRS:
            user_path = self.home_dir / skill_dir
            paths.append((user_path, "user"))

        return paths

    def discover_skills(self, skills_dir: Path) -> List[Path]:
        """
        在一个技能目录里获取全部的技能目录.

        一个技能目录必须包含一个SKILL.md文件 A skill directory must contain a SKILL.md file.

        Args:
            skills_dir: 包含有技能子目录的目录

        Returns:
            SKILL.md路径列表
        """

        # 如果技能目录不存在且当前不是目录，直接返回
        if not skills_dir.exists() or not skills_dir.is_dir():
            return []

        skill_files = []

        for item in skills_dir.iterdir():
            if item.is_dir():
                # 一个目录里面必须有一个SKILL.md
                skill_md = item / self.SKILL_FILE

                if skill_md.exists(): # md文件存在，则添加到md路径列表中
                    skill_files.append(skill_md)
                else:
                    # md不存在，则查找嵌套子目录
                    for nested_item in item.iterdir():
                        if nested_item.is_dir():
                            nested_skill_md = nested_item / self.SKILL_FILE
                            if nested_skill_md.exists():
                                skill_files.append(nested_skill_md)

        return skill_files

    def load_skill(self, skill_md_path: Path, location: str) -> Optional[Skill]:
        """
        从SKILL.md文件里加载一个单独的技能

        Args:
            skill_md_path: SKILL.md文件的路径
            location: Location type (project, user, managed)

        Returns:
            返回技能Skill实例，如果加载失败则返回None
        """
        try:
            skill = Skill.from_skill_md(skill_md_path, location)
            if skill:
                logger.debug(f"Loaded skill: {skill.name} from {skill_md_path}")
            else:
                logger.warning(f"Failed to parse skill: {skill_md_path}")
            return skill
        except Exception as e:
            logger.error(f"Error loading skill from {skill_md_path}: {e}")
            return None

    def load_all(self, registry: Optional[SkillRegistry] = None) -> SkillRegistry:
        """
        加载所有搜索路径中的技能，并注册到registry中

        Args:
            registry: Optional类型的registry，如果不存在则创建
.
        Returns:
            包含了全部技能的SkillRegistry
        """

        # 获取全局技能注册器
        if registry is None:
            registry = get_skill_registry()

        for search_path, location in self.get_search_paths():
            skill_files = self.discover_skills(search_path)

            for skill_md_path in skill_files:
                # 加载单个技能
                skill = self.load_skill(skill_md_path, location)
                if skill:
                    registered = registry.register(skill)
                    if registered:
                        logger.info(f"Registered skill: {skill.name} ({location})")
                    else:
                        logger.debug(f"Skipped skill {skill.name} - already registered from higher priority location")

        return registry

    def reload(self, registry: Optional[SkillRegistry] = None) -> SkillRegistry:
        """
        重新加载全部技能（先清空现有的技能）

        Args:
            registry: Optional registry to reload. Uses global if not provided.

        Returns:
            SkillRegistry containing all reloaded skills
        """
        if registry is None:
            registry = get_skill_registry()

        registry.clear()
        return self.load_all(registry)


def load_skills(project_root: Optional[Path] = None) -> SkillRegistry:
    """
    Convenience function to load all skills.

    Args:
        project_root: Root directory of the project

    Returns:
        SkillRegistry containing all loaded skills
    """
    loader = SkillLoader(project_root)
    return loader.load_all()


def get_available_skills() -> List[Skill]:
    """
    Get list of all available skills.

    Loads skills if the registry is empty.

    Returns:
        List of available skills
    """
    registry = get_skill_registry()

    if len(registry) == 0:
        load_skills()

    return registry.list_all()
